@alex I just discovered the inconsistency between boolean_mask() and its corresponding numpy operation.  For example, the following test code 
```
    foo_tf = tf.constant([[1, 2, 3, 4], [4, 5, 6, 7], [7, 8, 9, 10]])
    foo_np = np.array([[1, 2, 3, 4], [4, 5, 6, 7], [7, 8, 9, 10]])
    foo_mask = np.array([0.1, 1.1, 2.1], dtype=float)
    print(tf.boolean_mask(foo_tf, foo_mask).eval())
    print(foo_np[foo_mask])
```
will output
```
[[ 1  2  3  4]
 [ 4  5  6  7]
 [ 7  8  9 10]]
Traceback (most recent call last):
  File "test.py", line 46, in <module>
    print(foo_np[foo_mask])
IndexError: arrays used as indices must be of integer (or boolean) type
```

Is this a bug in boolean_mask() or is this result expected?
If it is expected, then I can keep writing test code. Otherwise, we need to first fix the bug existed in boolean_mask().

--- a/tensorflow/python/kernel_tests/array_ops_test.py
+++ b/tensorflow/python/kernel_tests/array_ops_test.py
@@ -514,17 +514,12 @@ class StridedSliceChecker(object):
       except AttributeError:
         return x
 
-    def convert_if_possible(spec):
-      try:
-        return ops.convert_to_tensor(spec)
-      except:
-        pass
-      return None
-
-    bool_spec = convert_if_possible(spec)
-    if bool_spec is not None and bool_spec.dtype == dtypes.bool:
+    if isinstance(spec, bool) or \
+      (isinstance(spec, ops.Tensor) and spec.dtype == dtypes.bool) or \
+      (isinstance(spec, np.ndarray) and spec.dtype == bool) or \
+      (isinstance(spec, (list, tuple)) and np.asarray(spec).dtype == bool):
       tensor = op.eval()
-      np_spec = eval_if_tensor(bool_spec)
+      np_spec = eval_if_tensor(spec)
       self.test.assertAllEqual(self.x_np[np_spec], tensor)
       return tensor
 
@@ -752,7 +747,7 @@ class StridedSliceTest(test_util.TensorFlowTestCase):
 
       # Test numpy array type mask
       raw = np.array([[[[[1, 2, 4, 5], [5, 6, 7, 8], [9, 10, 11, 12]]],
-        [[[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]]]]])
+                       [[[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]]]]])
       checker1 = StridedSliceChecker(self, raw)
       _ = checker1[raw >= 4]
       _ = checker1[raw < 19]

--- a/tensorflow/python/kernel_tests/array_ops_test.py
+++ b/tensorflow/python/kernel_tests/array_ops_test.py
@@ -514,13 +514,19 @@ class StridedSliceChecker(object):
       except AttributeError:
         return x
 
-    if isinstance(spec, (np.ndarray, ops.Tensor)):
+    def convert_if_possible(spec):
+      try:
+        return ops.convert_to_tensor(spec)
+      except:
+        pass
+      return None
+
+    bool_spec = convert_if_possible(spec)
+    if bool_spec is not None and bool_spec.dtype == dtypes.bool:
       tensor = op.eval()
-      np_specs = eval_if_tensor(spec)
-      ndims_np_specs = np_specs.ndim
-      if ndims_np_specs != 0:
-        self.test.assertAllEqual(self.x_np[np_specs], tensor)
-        return tensor
+      np_spec = eval_if_tensor(bool_spec)
+      self.test.assertAllEqual(self.x_np[np_spec], tensor)
+      return tensor
 
     if not isinstance(spec, (list, tuple)):
       spec = [spec]
@@ -687,6 +693,10 @@ class StridedSliceTest(test_util.TensorFlowTestCase):
         _ = checker[0.0]
       with self.assertRaisesRegexp(TypeError, expected):
         _ = checker[constant_op.constant(0.0)]
+      with self.assertRaisesRegexp(TypeError, expected):
+        _ = checker[constant_op.constant([1, 2, 3])]
+      with self.assertRaisesRegexp(TypeError, expected):
+        _ = checker[[2.1, -0.7, 1.5]]
 
   @test_util.run_deprecated_v1
   def testExpand(self):
@@ -747,6 +757,13 @@ class StridedSliceTest(test_util.TensorFlowTestCase):
       _ = checker1[raw >= 4]
       _ = checker1[raw < 19]
       _ = checker1[scalar]
+
+      # Test boolean and non boolean cases
+      mask = np.array([True, False, True])
+      raw1 = np.array([[1, 2, 4, 5], [5, 6, 7, 8], [9, 10, 11, 12]])
+      checker2 = StridedSliceChecker(self, raw1)
+      _ = checker2[mask]
+      _ = checker2[ops.convert_to_tensor(mask)]
 
 
 class StridedSliceShapeChecker(object):

--- a/tensorflow/python/kernel_tests/array_ops_test.py
+++ b/tensorflow/python/kernel_tests/array_ops_test.py
@@ -747,7 +747,8 @@ class StridedSliceTest(test_util.TensorFlowTestCase):
 
       # Test numpy array type mask
       raw = np.array([[[[[1, 2, 4, 5], [5, 6, 7, 8], [9, 10, 11, 12]]],
-                       [[[13, 14, 15, 16], [17, 18, 19, 20], [21, 22, 23, 24]]]]])
+                       [[[13, 14, 15, 16], [17, 18, 19, 20],
+                         [21, 22, 23, 24]]]]])
       checker1 = StridedSliceChecker(self, raw)
       _ = checker1[raw >= 4]
       _ = checker1[raw < 19]
@@ -860,7 +861,7 @@ class GradSliceChecker(object):
     np_val_grad = (2 * self.varnp * self.varnp)
     np_sliceval_grad = np.zeros(self.var.get_shape())
     if isinstance(spec, ops.Tensor):
-        spec = self.sess.run([spec])
+      spec = self.sess.run([spec])
     np_sliceval_grad[spec] = np_val_grad[spec]
     # verify gradient
     self.test.assertAllEqual(slice_val_grad_evaled, np_sliceval_grad)

--- a/tensorflow/python/ops/array_ops.py
+++ b/tensorflow/python/ops/array_ops.py
@@ -579,16 +579,16 @@ def _slice_helper(tensor, slice_spec, var=None):
     TypeError: If the slice indices aren't int, slice, ellipsis,
       tf.newaxis or scalar int32/int64 tensors.
   """
-  if isinstance(slice_spec, np.ndarray):
-    ndims_slice_spec = slice_spec.ndim
-    if ndims_slice_spec != 0:
-      return boolean_mask(tensor=tensor, mask=slice_spec)
+  def convert_if_possible(slice_spec):
+    try:
+      return ops.convert_to_tensor(slice_spec)
+    except:
+      pass
+    return None
 
-  if isinstance(slice_spec, ops.Tensor):
-    shape_slice_spec = slice_spec.get_shape()
-    ndims_slice_spec = shape_slice_spec.ndims
-    if ndims_slice_spec != 0:
-      return boolean_mask(tensor=tensor, mask=slice_spec)
+  bool_spec = convert_if_possible(slice_spec)
+  if bool_spec is not None and bool_spec.dtype == dtypes.bool:
+    return boolean_mask(tensor=tensor, mask=bool_spec)
 
   begin, end, strides = [], [], []
   index = 0

--- a/tensorflow/python/ops/array_ops.py
+++ b/tensorflow/python/ops/array_ops.py
@@ -580,9 +580,10 @@ def _slice_helper(tensor, slice_spec, var=None):
       tf.newaxis or scalar int32/int64 tensors.
   """
   if isinstance(slice_spec, bool) or \
-    (isinstance(slice_spec, ops.Tensor) and slice_spec.dtype == dtypes.bool) or \
-    (isinstance(slice_spec, np.ndarray) and slice_spec.dtype == bool) or \
-    (isinstance(slice_spec, (list, tuple)) and np.asarray(slice_spec).dtype == bool):
+  (isinstance(slice_spec, ops.Tensor) and slice_spec.dtype == dtypes.bool) or \
+  (isinstance(slice_spec, np.ndarray) and slice_spec.dtype == bool) or \
+  (isinstance(slice_spec, (list, tuple)) and
+   np.asarray(slice_spec).dtype == bool):
     return boolean_mask(tensor=tensor, mask=slice_spec)
 
   if not isinstance(slice_spec, (list, tuple)):
@@ -1333,11 +1334,11 @@ def boolean_mask(tensor, mask, name="boolean_mask", axis=None):
     axis = 0 if axis is None else axis
     shape_tensor[axis:axis + ndims_mask].assert_is_compatible_with(shape_mask)
 
-    leading_size = gen_math_ops.prod(shape_tensor[axis:axis + ndims_mask], [0])
+    leading_size = gen_math_ops.prod(shape(tensor)[axis:axis + ndims_mask], [0])
     tensor = reshape(tensor,
                      concat([
-                         shape_tensor[:axis], [leading_size],
-                         shape_tensor[axis + ndims_mask:]
+                         shape(tensor)[:axis], [leading_size],
+                         shape(tensor)[axis + ndims_mask:]
                      ], 0))
     first_dim = shape_tensor[axis:axis + ndims_mask].num_elements()
     tensor.set_shape(

--- a/tensorflow/python/ops/array_ops.py
+++ b/tensorflow/python/ops/array_ops.py
@@ -579,16 +579,14 @@ def _slice_helper(tensor, slice_spec, var=None):
     TypeError: If the slice indices aren't int, slice, ellipsis,
       tf.newaxis or scalar int32/int64 tensors.
   """
-  def convert_if_possible(slice_spec):
-    try:
-      return ops.convert_to_tensor(slice_spec)
-    except:
-      pass
-    return None
+  if isinstance(slice_spec, bool) or \
+    (isinstance(slice_spec, ops.Tensor) and slice_spec.dtype == dtypes.bool) or \
+    (isinstance(slice_spec, np.ndarray) and slice_spec.dtype == bool) or \
+    (isinstance(slice_spec, (list, tuple)) and np.asarray(slice_spec).dtype == bool):
+    return boolean_mask(tensor=tensor, mask=slice_spec)
 
-  bool_spec = convert_if_possible(slice_spec)
-  if bool_spec is not None and bool_spec.dtype == dtypes.bool:
-    return boolean_mask(tensor=tensor, mask=bool_spec)
+  if not isinstance(slice_spec, (list, tuple)):
+    slice_spec = [slice_spec]
 
   begin, end, strides = [], [], []
   index = 0

